### ISSUE TITLE
UnicodeDecodeError: 'utf-8' codec can't decode byte 0xa0 in position 16800744: invalid start byte

### ISSUE THREAD

**Issue by butterflyhigh** [mesonbuild/meson#13443]
**Describe the bug**
When I try to run `meson setup` while chrooted into a minimal Linux system, it fails due to the unicode error in the title. It does not fail on the host system.

The output of `meson setup` from chroot:
```
The Meson build system
Version: 1.5.0
Source dir: /home/portage
Build dir: /home/portage/build
Build type: native build
Project name: portage
Project version: 3.0.65
C compiler for the host machine: clang (clang 18.1.8 "clang version 18.1.8")
C linker for the host machine: clang ld.bfd 2.42.0
Host machine cpu family: x86_64
Host machine cpu: x86_64
Program python3 found: YES (/usr/bin/python)
Program sed found: YES (/usr/bin/sed)
Configuring const.py using configuration
Configuring installation.py using configuration
Configuring __init__.py using configuration
Configuring AbstractEbuildProcess.py using configuration
Configuring create_depgraph_params.py using configuration
Found pkg-config: YES (/usr/bin/pkg-config) 2.1.1
Run-time dependency python found: YES 3.12
Configuring METADATA using configuration
Program patch found: YES (/usr/bin/patch)
Configuring make.conf.example with command
Program sphinx-apidoc found: NO
Program sphinx-build found: NO
Configuring version using configuration
Program xmlto found: NO
Configuring color.map.5 using configuration
Configuring dispatch-conf.1 using configuration
Configuring ebuild.1 using configuration
Configuring ebuild.5 using configuration
Configuring egencache.1 using configuration
Configuring emaint.1 using configuration
Configuring emerge.1 using configuration
Configuring emirrordist.1 using configuration
Configuring env-update.1 using configuration
Configuring etc-update.1 using configuration
Configuring fixpackages.1 using configuration
Configuring glsa-check.1 using configuration
Configuring make.conf.5 using configuration
Configuring portage.5 using configuration
Configuring quickpkg.1 using configuration
Configuring xpak.5 using configuration
Configuring color.map.5 using configuration
Configuring dispatch-conf.1 using configuration
Configuring ebuild.1 using configuration
Configuring env-update.1 using configuration
Configuring etc-update.1 using configuration
Configuring fixpackages.1 using configuration
Build targets in project: 1

Found ninja-1.12.1 at /usr/bin/ninja
Traceback (most recent call last):                                                                                                                                                                                
  File "/usr/lib/python3.12/site-packages/mesonbuild/mesonmain.py", line 188, in run
    return options.run_func(options)
           ^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3.12/site-packages/mesonbuild/msetup.py", line 364, in run
    app.generate()
  File "/usr/lib/python3.12/site-packages/mesonbuild/msetup.py", line 187, in generate
    return self._generate(env, capture, vslite_ctx)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3.12/site-packages/mesonbuild/msetup.py", line 252, in _generate
    captured_compile_args = intr.backend.generate(capture, vslite_ctx)
                            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3.12/site-packages/mesonbuild/backend/ninjabackend.py", line 648, in generate
    self.generate_target(t)
  File "/usr/lib/python3.12/site-packages/mesonbuild/backend/ninjabackend.py", line 1048, in generate_target
    o, s = self.generate_single_compile(target, src, False, [],
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3.12/site-packages/mesonbuild/backend/ninjabackend.py", line 2984, in generate_single_compile
    self.create_target_source_introspection(target, compiler, commands, [src], [], unity_sources)
  File "/usr/lib/python3.12/site-packages/mesonbuild/backend/ninjabackend.py", line 822, in create_target_source_introspection
    parameters = parameters.to_native(copy=True)
                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3.12/site-packages/mesonbuild/compilers/mixins/clike.py", line 92, in to_native
    default_dirs = self.compiler.get_default_include_dirs()
                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3.12/site-packages/mesonbuild/compilers/mixins/gnu.py", line 402, in get_default_include_dirs
    return gnulike_default_include_dirs(tuple(self.get_exelist(ccache=False)), self.language).copy()
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3.12/site-packages/mesonbuild/compilers/mixins/gnu.py", line 327, in gnulike_default_include_dirs
    _, stdout, _ = mesonlib.Popen_safe(cmd, stderr=subprocess.STDOUT, env=env)
                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3.12/site-packages/mesonbuild/utils/universal.py", line 1554, in Popen_safe
    o, e = p.communicate(write)
           ^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3.12/subprocess.py", line 1196, in communicate
    stdout = self.stdout.read()
             ^^^^^^^^^^^^^^^^^^
  File "<frozen codecs>", line 322, in decode
UnicodeDecodeError: 'utf-8' codec can't decode byte 0xa0 in position 16800744: invalid start byte

ERROR: Unhandled python exception

    This is a Meson bug and should be reported!
```

The output of `meson setup` from the host system (Arch Linux):
```
The Meson build system
Version: 1.5.0
Source dir: /home/coolkid/portage
Build dir: /home/coolkid/portage/build
Build type: native build
Project name: portage
Project version: 3.0.65
C compiler for the host machine: ccache cc (gcc 14.1.1 "cc (GCC) 14.1.1 20240522")
C linker for the host machine: cc ld.bfd 2.42.0
Host machine cpu family: x86_64
Host machine cpu: x86_64
Program python3 found: YES (/usr/bin/python)
Program sed found: YES (/usr/bin/sed)
Configuring const.py using configuration
Configuring installation.py using configuration
Configuring __init__.py using configuration
Configuring AbstractEbuildProcess.py using configuration
Configuring create_depgraph_params.py using configuration
Found pkg-config: YES (/usr/bin/pkg-config) 2.1.1
Run-time dependency python found: YES 3.12
Configuring METADATA using configuration
Program patch found: YES (/usr/bin/patch)
Configuring make.conf.example with command
Program sphinx-apidoc found: NO
Program sphinx-build found: NO
Configuring version using configuration
Program xmlto found: NO
Configuring color.map.5 using configuration
Configuring dispatch-conf.1 using configuration
Configuring ebuild.1 using configuration
Configuring ebuild.5 using configuration
Configuring egencache.1 using configuration
Configuring emaint.1 using configuration
Configuring emerge.1 using configuration
Configuring emirrordist.1 using configuration
Configuring env-update.1 using configuration
Configuring etc-update.1 using configuration
Configuring fixpackages.1 using configuration
Configuring glsa-check.1 using configuration
Configuring make.conf.5 using configuration
Configuring portage.5 using configuration
Configuring quickpkg.1 using configuration
Configuring xpak.5 using configuration
Configuring color.map.5 using configuration
Configuring dispatch-conf.1 using configuration
Configuring ebuild.1 using configuration
Configuring env-update.1 using configuration
Configuring etc-update.1 using configuration
Configuring fixpackages.1 using configuration
Build targets in project: 1

Found ninja-1.12.1 at /usr/bin/ninja
```

**To Reproduce**
[The repository being built](https://github.com/gentoo/portage)

**Expected behavior**
I expected it to build successfully. 

**system parameters**
* Is this a [cross build](https://mesonbuild.com/Cross-compilation.html) or just a plain native build (for the same computer)? No, it is on a different drive but sharing the same CPU
* what operating system (e.g. MacOS Catalina, Windows 10, CentOS 8.0, Ubuntu 18.04, etc.)? The problematic system was created by `pacstrap`ing the Linux kernel and several useful programs, but without pacman or other arch-specific programs. `uname -srm` outputs `Linux 6.9.9-arch1-1 x86_64`
* what Python version are you using e.g. 3.8.0? Python 3.12.4
* what `meson --version`? 1.5.0
* what `ninja --version` if it's a Ninja build? 1.12.1

**Comment by eli-schwartz** [comment]
The error message comes from here: https://github.com/mesonbuild/meson/blob/864919973f73ec2c6ae6ca01be7fe5e8b950df12/mesonbuild/compilers/mixins/gnu.py#L320-L327


Inside the chroot, you have clang 18, and it attempts to:
- export `LC_ALL=C` into the subprocess
- fork and run: `clang -xc -E -v -`
- close stdin and read stdout+stderr

This apparently fails because the combination of stdout and stderr for clang when told to use the C locale, is... not utf-8? This does not seem like a thing that should happen.

**Comment by rgonzalezfluendo** [comment]
I could reproduce this issue.

Using a wrapper over `gcc`, we found no empty stdin and stdout/stderr is a not valid utf-8:
```	Running: gcc -xc -E -v -
Using built-in specs.
COLLECT_GCC=gcc2
OFFLOAD_TARGET_NAMES=nvptx-none:hsa
OFFLOAD_TARGET_DEFAULT=1
Target: x86_64-linux-gnu
Configured with: ../src/configure -v --with-pkgversion='Ubuntu 9.4.0-1ubuntu1~20.04.2' --with-bugurl=file:///usr/share/doc/gcc-9/README.Bugs --enable-languages=c,ada,c++,go,brig,d,fortran,objc,obj-c++,gm2 --prefix=/usr --with-gcc-major-version-only --program-suffix=-9 --program-prefix=x86_64-linux-gnu- --enable-shared --enable-linker-build-id --libexecdir=/usr/lib --without-included-gettext --enable-threads=posix --libdir=/usr/lib --enable-nls --enable-clocale=gnu --enable-libstdcxx-debug --enable-libstdcxx-time=yes --with-default-libstdcxx-abi=new --enable-gnu-unique-object --disable-vtable-verify --enable-plugin --enable-default-pie --with-system-zlib --with-target-system-zlib=auto --enable-objc-gc=auto --enable-multiarch --disable-werror --with-arch-32=i686 --with-abi=m64 --with-multilib-list=m32,m64,mx32 --enable-multilib --with-tune=generic --enable-offload-targets=nvptx-none=/build/gcc-9-9QDOt0/gcc-9-9.4.0/debian/tmp-nvptx/usr,hsa --without-cuda-driver --enable-checking=release --build=x86_64-linux-gnu --host=x86_64-linux-gnu --target=x86_64-linux-gnu
Thread model: posix
gcc version 9.4.0 (Ubuntu 9.4.0-1ubuntu1~20.04.2)
COLLECT_GCC_OPTIONS='-E' '-v' '-mtune=generic' '-march=x86-64'
 /usr/lib/gcc/x86_64-linux-gnu/9/cc1 -E -quiet -v -imultiarch x86_64-linux-gnu - -mtune=generic -march=x86-64 -fasynchronous-unwind-tables -fstack-protector-strong -Wformat -Wformat-security -fstack-clash-protection -fcf-protection
ignoring nonexistent directory "/usr/local/include/x86_64-linux-gnu"
ignoring nonexistent directory "/usr/lib/gcc/x86_64-linux-gnu/9/include-fixed"
ignoring nonexistent directory "/usr/lib/gcc/x86_64-linux-gnu/9/../../../../x86_64-linux-gnu/include"
#include "..." search starts here:
#include <...> search starts here:
 /opt/codecs-gst-latest-linux-x86_64/dist/linux_x86_64/include
 /usr/lib/gcc/x86_64-linux-gnu/9/include
 /usr/local/include
 /usr/include/x86_64-linux-gnu
 /usr/include
End of search list.
<stdin>:2:1: warning: null character(s) ignored
<stdin>:2:52: warning: null character(s) ignored
<stdin>:2:63: warning: null character(s) ignored
<stdin>:2:72: warning: null character(s) ignored
<stdin>:2:87: warning: null character(s) ignored
<stdin>:2:120: warning: null character(s) ignored
<stdin>:2:239: warning: null character(s) ignored
<stdin>:2:261: warning: null character(s) ignored
<stdin>:2:273: warning: null character(s) ignored
<stdin>:2:295: warning: null character(s) ignored
<stdin>:2:328: warning: null character(s) ignored
<stdin>:3:1: warning: null character(s) ignored
<stdin>:3:10: warning: null character(s) ignored
<stdin>:3:19: warning: null character(s) ignored
<stdin>:3:34: warning: null character(s) ignored
<stdin>:3:63: warning: null character(s) ignored
<stdin>:3:182: warning: null character(s) ignored
<stdin>:3:204: warning: null character(s) ignored
<stdin>:3:216: warning: null character(s) ignored
<stdin>:3:238: warning: null character(s) ignored
<stdin>:3:267: warning: null character(s) ignored
# 1 "<stdin>"
# 1 "<built-in>"
# 1 "<command-line>"
# 31 "<command-line>"
# 1 "/usr/include/stdc-predef.h" 1 3 4
# 32 "<command-line>" 2
# 1 "<stdin>"
No LSB modules are available.
                                                  ^P g\361Dh ^A...
```

**Comment by rgonzalezfluendo** [comment]
eureka!!!

The issue is about running Docker as root and using `/dev/null` for storing mess.


TL;DR: 
```
root@d451f47dc3df:~# echo "123" > /dev/null.tmp5E12O3 && mv /dev/null.tmp5E12O3 /dev/null && cat /dev/null
123
```

In my case we were using `/dev/null` to write the GStreamer registry, which is a binary that causes the failure when running `gcc -xc -E -v -`.

It't the same as running `<~/.cache/gstreamer-1.0/registry.x86_64.bin gcc -xc -E -v -`

<details>
 <summary>Simple code to reproduce the issue inside docker as root</summary>

```py
import os
import subprocess

from mesonbuild import mesonlib


env = os.environ.copy()
env["GST_REGISTRY_1_0"] = '/dev/null'

# Run the command and capture output
# "apt install gstreamer1.0-tools" needed
result = subprocess.run(
    ['gst-inspect-1.0'],
    capture_output=True,
    text=True,
    check=True,
    env=env
)
print(result)

compiler = ["gcc"]
lang = "c"

env = os.environ.copy()
env["LC_ALL"] = 'C'
cmd = list(compiler) + [f'-x{lang}', '-E', '-v', '-']
print(cmd)
_, stdout, _ = mesonlib.Popen_safe(cmd, stderr=subprocess.STDOUT, env=env)
print(stdout)
```

</details>

**Comment by QuLogic** [comment]
> ```
> echo "123" > /dev/null.tmp5E12O3 && mv /dev/null.tmp5E12O3 /dev/null && cat /dev/null
> ```

This makes no sense; you've replaced `/dev/null` with something that doesn't act the same as `/dev/null`, so of course things are broken?

**Comment by rgonzalezfluendo** [comment]
100%. Replacing `/dev/null` is nonsense. We were doing it within a test without being aware of the implications. I'm sharing this here so whoever created the issue can see if it's the same issue. No fix needed at meson level.

**Comment by dcbaker** [comment]
I'm going to close this as resolved then. We can re-open if the this can be replicated without replacing /dev/null
